gateway/transport/plugins/audit/wal.go

```diff
@@ -82,11 +83,65 @@ func (p *auditPlugin) dropWalLog(sid string) {
 }
 
 func (p *auditPlugin) writeOnClose(pctx plugintypes.Context, errMsg error) error {
-	walLogObj := p.walSessionStore.Pop(pctx.SID)
-	walogm, ok := walLogObj.(*walLogRWMutex)
+	walFolder := fmt.Sprintf(walFolderTmpl, plugintypes.AuditPath, pctx.OrgID, pctx.SID)
+	logsFileExist := wal.FileExists(walFolder)
+	if !logsFileExist {
+		// if the wal file does not exist we neet to update the session event stream
+		blobStream := fmt.Sprintf(`[ [1.0, "%s", "%s"] ]`, "e", base64.StdEncoding.EncodeToString([]byte("no log found on disk")))
+		emptyMetrics := make(map[string]any, 0)
+
+		var blobFormat *string
+		switch pctx.ProtoConnectionType() {
+		case pb.ConnectionTypePostgres:
+			blobFormat = ptr.String(models.BlobFormatWireProtoType)
+		}
+
+		endDate := time.Now().UTC()
+		err := models.UpdateSessionEventStream(models.SessionDone{
+			ID:         pctx.SID,
+			OrgID:      pctx.OrgID,
+			Metrics:    emptyMetrics,
+			BlobStream: json.RawMessage(blobStream),
+			BlobFormat: blobFormat,
+			Status:     string(openapi.SessionStatusDone),
+			ExitCode:   parseExitCodeFromErr(errMsg),
+			EndSession: &endDate,
+		})
+		log.With("sid", pctx.SID, "origin", pctx.ClientOrigin, "verb", pctx.ClientVerb).
+			Infof("finished persisting session to store, update-session-err=%v, context-err=%v", err, errMsg)
+
+		if err != nil {
+			log.With("sid", pctx.SID).Warnf("failed updating session event stream: %v", err)
+		} else {
+			if err := os.RemoveAll(walFolder); err != nil {
+				log.Errorf("failed removing wal file %q, err=%v", walFolder, err)
+			}
+		}
+		log.With("sid", pctx.SID).Debugf("no wal log found on disk for session, path=%v, err=%v", walFolder, err)
+		p.walSessionStore.Pop(pctx.SID)
+		return err
+
+	}
+	// first we gonna try to obtain the wal from memory because could have unflushed logs from memory to disk
+	walLogObjMemory := p.walSessionStore.Pop(pctx.SID)
```

Comment:
I am trying to use the memory first, because if it is in the middle of a kill session by the user, it might not have been written to the disk yet, so getting from the disk will lose the last logs, something... if the memory is not present anymore, I will go for the disk.